### PR TITLE
HADOOP-16090 S3A Client to add explicit support for versioned stores.

### PR DESCRIPTION
**Pull request by steveloughran**
Adds the option fs.s3a.versioned.store to enable all policy changes for efficiently working with versioned object stores.

This (initial?) patch reverts to the original probe-then-delete mechanism of cleaning up fake empty directories; the test ITestS3AFileOperationCost.testFakeDirectoryDeletion() verifies that this
reduces the number of paths queued for deletion.

Change-Id: I9708ffc3784e8f7e742b5885b48cbaf9600f6232

### CODE REVIEW

**Comment by steveloughran** [comment]
Tested: the ITestS3AFileOperationCost test run against S3 london (versioned); yet to rerun full suite

File: hadoop-tools/hadoop-aws/src/site/markdown/tools/hadoop-aws/index.md [hunk 1]
```diff
@@ -1270,6 +1270,46 @@ will be thrown.  When `false` and and eTag or version ID is not returned,
...
+  <value>false</value>
+  <description>
+    Is the S3 Bucket "versioned"? If so, the S3A connector
+    will optimize its behavior for versioned data.
+  </description>
+</property>
+```
+
+Setting this option for non-versioned stores is not dangerous, but likely
+to make some operations slower.
```
Comment (hadoop-yetus) [inline hadoop-tools/hadoop-aws/src/site/markdown/tools/hadoop-aws/index.md]: whitespace:end of line

File: hadoop-tools/hadoop-aws/src/site/markdown/tools/hadoop-aws/index.md [hunk 2]
```diff
@@ -1270,6 +1270,46 @@ will be thrown.  When `false` and and eTag or version ID is not returned,
...
+  <value>versionid</value>
+</property>
+```
+
+Note: the specific optimizations which the S3A client does for versioned
+object stores may change across versions: the goals are
+
+* Reduce the number of needless key deletion requests made, even at the expense of more
+HEAD operations [HADOOP-16090](https://issues.apache.org/jira/browse/HADOOP-16090).
+* Use version IDs where possible.
```
Comment (hadoop-yetus) [inline hadoop-tools/hadoop-aws/src/site/markdown/tools/hadoop-aws/index.md]: whitespace:end of line

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 25 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 1 new or modified test files. |
||| _ trunk Compile Tests _ |
| 0 | mvndep | 24 | Maven dependency ordering for branch |
| +1 | mvninstall | 988 | trunk passed |
| +1 | compile | 932 | trunk passed |
| +1 | checkstyle | 231 | trunk passed |
| +1 | mvnsite | 138 | trunk passed |
| +1 | shadedclient | 1088 | branch has no errors when building and testing our client artifacts. |
| +1 | findbugs | 159 | trunk passed |
| +1 | javadoc | 103 | trunk passed |
||| _ Patch Compile Tests _ |
| 0 | mvndep | 24 | Maven dependency ordering for patch |
| +1 | mvninstall | 76 | the patch passed |
| +1 | compile | 902 | the patch passed |
| +1 | javac | 902 | the patch passed |
| -0 | checkstyle | 191 | root: The patch generated 3 new + 5 unchanged - 0 fixed = 8 total (was 5) |
| +1 | mvnsite | 124 | the patch passed |
| -1 | whitespace | 0 | The patch has 2 line(s) that end in whitespace. Use git apply --whitespace=fix <<patch_file>>. Refer https://git-scm.com/docs/git-apply |
| +1 | xml | 1 | The patch has no ill-formed XML file. |
| +1 | shadedclient | 650 | patch has no errors when building and testing our client artifacts. |
| +1 | findbugs | 164 | the patch passed |
| +1 | javadoc | 97 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 497 | hadoop-common in the patch passed. |
| +1 | unit | 292 | hadoop-aws in the patch passed. |
| +1 | asflicense | 49 | The patch does not generate ASF License warnings. |
| | | 6671 | |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=17.05.0-ce Server=17.05.0-ce base: https://builds.apache.org/job/hadoop-multibranch/job/PR-621/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/621 |
| Optional Tests |  dupname  asflicense  compile  javac  javadoc  mvninstall  mvnsite  unit  shadedclient  xml  findbugs  checkstyle  |
| uname | Linux d5e8e8a6ad17 4.4.0-138-generic #164-Ubuntu SMP Tue Oct 2 17:16:02 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / ae3a2c3 |
| maven | version: Apache Maven 3.3.9 |
| Default Java | 1.8.0_191 |
| findbugs | v3.1.0-RC1 |
| checkstyle | https://builds.apache.org/job/hadoop-multibranch/job/PR-621/1/artifact/out/diff-checkstyle-root.txt |
| whitespace | https://builds.apache.org/job/hadoop-multibranch/job/PR-621/1/artifact/out/whitespace-eol.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-621/1/testReport/ |
| Max. process+thread count | 1573 (vs. ulimit of 5500) |
| modules | C: hadoop-common-project/hadoop-common hadoop-tools/hadoop-aws U: . |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-621/1/console |
| Powered by | Apache Yetus 0.9.0 http://yetus.apache.org |


This message was automatically generated.

**Comment by steveloughran** [comment]
On the ASF JIRA, it's being reported that testing this for Flink checkpointing is generating enough HEAD requests that writes are being throttled. DELETE calls aren't throttled, see, whereas the poll & delete operation was doing 1 HEAD/entry.

I wonder if we could be more minimal and only look for a fake directory marker in the parent, so 1 HEAD per write only.

File: hadoop-tools/hadoop-aws/src/site/markdown/tools/hadoop-aws/index.md (hunk 1 above)
Comment (hadoop-yetus) [inline hadoop-tools/hadoop-aws/src/site/markdown/tools/hadoop-aws/index.md]: whitespace:end of line

File: hadoop-tools/hadoop-aws/src/site/markdown/tools/hadoop-aws/index.md (hunk 2 above)
Comment (hadoop-yetus) [inline hadoop-tools/hadoop-aws/src/site/markdown/tools/hadoop-aws/index.md]: whitespace:end of line

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 27 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 1 new or modified test files. |
||| _ trunk Compile Tests _ |
| 0 | mvndep | 24 | Maven dependency ordering for branch |
| +1 | mvninstall | 1021 | trunk passed |
| +1 | compile | 974 | trunk passed |
| +1 | checkstyle | 189 | trunk passed |
| +1 | mvnsite | 119 | trunk passed |
| +1 | shadedclient | 1004 | branch has no errors when building and testing our client artifacts. |
| +1 | findbugs | 148 | trunk passed |
| +1 | javadoc | 86 | trunk passed |
||| _ Patch Compile Tests _ |
| 0 | mvndep | 23 | Maven dependency ordering for patch |
| +1 | mvninstall | 77 | the patch passed |
| +1 | compile | 971 | the patch passed |
| +1 | javac | 971 | the patch passed |
| -0 | checkstyle | 185 | root: The patch generated 3 new + 5 unchanged - 0 fixed = 8 total (was 5) |
| +1 | mvnsite | 110 | the patch passed |
| -1 | whitespace | 0 | The patch has 2 line(s) that end in whitespace. Use git apply --whitespace=fix <<patch_file>>. Refer https://git-scm.com/docs/git-apply |
| +1 | xml | 1 | The patch has no ill-formed XML file. |
| +1 | shadedclient | 664 | patch has no errors when building and testing our client artifacts. |
| +1 | findbugs | 173 | the patch passed |
| +1 | javadoc | 87 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 503 | hadoop-common in the patch passed. |
| +1 | unit | 287 | hadoop-aws in the patch passed. |
| +1 | asflicense | 44 | The patch does not generate ASF License warnings. |
| | | 6653 | |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=17.05.0-ce Server=17.05.0-ce base: https://builds.apache.org/job/hadoop-multibranch/job/PR-621/2/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/621 |
| Optional Tests |  dupname  asflicense  compile  javac  javadoc  mvninstall  mvnsite  unit  shadedclient  xml  findbugs  checkstyle  |
| uname | Linux 11e6d2943a01 4.4.0-138-generic #164-Ubuntu SMP Tue Oct 2 17:16:02 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / d2637cb |
| maven | version: Apache Maven 3.3.9 |
| Default Java | 1.8.0_191 |
| findbugs | v3.1.0-RC1 |
| checkstyle | https://builds.apache.org/job/hadoop-multibranch/job/PR-621/2/artifact/out/diff-checkstyle-root.txt |
| whitespace | https://builds.apache.org/job/hadoop-multibranch/job/PR-621/2/artifact/out/whitespace-eol.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-621/2/testReport/ |
| Max. process+thread count | 1573 (vs. ulimit of 5500) |
| modules | C: hadoop-common-project/hadoop-common hadoop-tools/hadoop-aws U: . |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-621/2/console |
| Powered by | Apache Yetus 0.9.0 http://yetus.apache.org |


This message was automatically generated.

**Pull request closed without merging**